### PR TITLE
[query] Evaluate relational lets in LowerToCDA.

### PR DESCRIPTION
**Pull request by tpoterba**
Up to 538 passing.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerToCDA.scala [hunk 1]
```diff
@@ -10,15 +15,56 @@ object LowerToCDA {
 
     // Slightly simpler to replace RelationalRefs in a second linear pass,
     // since we'd have to embed the rewrite in many places otherwise
-    RewriteBottomUp(lower(ir, typesToLower, ctx, r, Seq()), {
+    RewriteBottomUp(lower(ir, typesToLower, ctx, r, Map()), {
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerToCDA.scala:18]: Do you need the second pass any more? Doesn't `lower` now leave no `RelationalRef`s`

File: hail/src/main/scala/is/hail/expr/ir/LiftRelationalValues.scala [hunk 2]
```diff
@@ -7,6 +7,11 @@ object LiftRelationalValues {
   def apply(ir0: IR): IR = {
 
     def rewrite(ir: BaseIR, ab: ArrayBuilder[(String, IR)]): BaseIR = ir match {
+      case RelationalLet(name, value, body) =>
+        val value2 = rewrite(value, ab).asInstanceOf[IR]
+        val ab2 = new ArrayBuilder[(String, IR)]
+        val body2 = rewrite(body, ab2).asInstanceOf[IR]
+        RelationalLet(name, value2, ab2.result().foldRight[IR](body2) { case ((name, value), acc) => RelationalLet(name, value, acc) })
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LiftRelationalValues.scala:14]: no, that could lift references to `name` above the binding.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerToCDA.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerToCDA.scala:18]: oh, yeah, forgot to remove this.

**Comment by tpoterba** [comment]
Up to 555 tests passing against local backend with this.

File: hail/src/main/scala/is/hail/expr/ir/LiftRelationalValues.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/LiftRelationalValues.scala:14]: I don't see that. The call to `LiftRelationalValues(body)` expands to
```
      case RelationalLet(name, value, body) =>
        val value2 = rewrite(value, ab).asInstanceOf[IR]
        val body2 = {
          val ab = new ArrayBuilder[(String, IR)]
          val rw = rewrite(body, ab).asInstanceOf[IR]
          ab.result().foldRight[IR](rw) { case ((name, value), acc) => RelationalLet(name, value, acc)
        }
        RelationalLet(name, value2, body2)
```
which is equivalent to what you have. But it's a minor improvement, not worth holding up the PR.

[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LiftRelationalValues.scala:14]: sorry, I misread! you're totally right.

**Pull request merged**